0002-snowflake-ids.md

```diff
@@ -0,0 +1,71 @@
+
+Finally, an ID beginning with the timestamp can easily be sorted by, and provide a reference point to get messages before and after it.
+
+## Detailed design
+
+Snowflakes are stored in 64-bit unsigned integers. The client may only depend on the timestamp, as the other parts may be changed without warning.
+
+### Timestamp
+
+Bits 63 through 22 are used for the timestamp. The epoch for this timestamp is midnight, January 1st, 2023. The Unix timestamp for this epoch is 1672531200000. This means that to convert to Unix time, the timestamp of the epoch should be added, and vice versa. This has space for times up to May 15th, 2162.
+
+The timestamp of an ID may never be less than that of an ID generated earlier with the same worker ID. IDs across separate workers should be sufficiently close for the purpose of sorting, ideally less than 1 second.
+
+In cases where only signed integers are available, the first bit of the timestamp may be ignored by the client, to account for the sign bit.
+
+### Worker ID
+
+Bits 21 through 12 are used for the worker ID. This is an internal ID assigned to the specific process that generated the ID by the backend, and must be unique for each process that generates IDs. The manner in which this is assigned is unspecified.
+
+### Counter
+
+Bits 11 through 0 are used for the counter. The counter increments each time an ID is generated. It may never decrease, except for when the clock used for the timestamp ticks to the next millisecond, in which case it is reset to zero. The counter may never overflow. If it reaches the 12-bit limit of 4095 before the timestamp clock has ticked to the next millisecond, no more IDs may be returned until this happens.
+
+## Drawbacks
+
+- 64 bits may be considered too little space. It should however be possible to migrate to longer IDs later while preserving then-legacy ones.
+- The possibility of having federation in the future means that instances might end up sharing IDs, and this format does not identify the origin instance. This could mean a large change in how IDs work and/or are managed in case this is implemented.
+- Clients have to keep track of which instance an ID belongs to.
+
+## Alternatives
+
+What other solutions have been considered? Why this solution rather than them? What is the impact of not doing this?
+
+- A custom format was proposed, though it was discarded for being unproven and providing little benefit over snowflakes.
+- Including a hash of the origin instance's domain name was considered, assuming federation would be supported, though it was dismissed for its high risk of hash collisions, dealing with which would mean taking bits away from the other parts of the ID that provide more effective means of ID collisions.
+- UUIDv7 was proposed, though it does not appear to have much benefit over snowflakes, while being twice as long.
```

Comment:
Saying this while also saying:
> 64 bits may be considered too little space

Is a bit contradictory, though criticizing them for not having worker ids could be justifiable (though they have randomness reducing chances of this being a issue)

Reply:
I don't think it is. A lack of space is not a consideration I agree with, but it is a consideration nonetheless. Additionally, the sentence right after that second quote states:

> It should however be possible to migrate to longer IDs later while preserving then-legacy ones.

Meaning that if it becomes the case, it is solvable. Would this clarification help?

```suggestion
- UUIDv7 was proposed, though it does not appear to have much benefit over snowflakes, and its greater size does not seem necessary or useful at the moment. Later migration remains an option if needed.
```

Reply:
I'm good with it

Reply:
I'd just like to mention that in federation, hash collisions would not be any more impactful than snowflake collisions, and would in fact be more unlikely.

Reply:
I don't think anyone claimed otherwise. This is not about federated versus not federated.